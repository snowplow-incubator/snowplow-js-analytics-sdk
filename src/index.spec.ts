/*
 * Copyright (c) 2018-2019 dokmic, Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

import { transform } from './index';
import { event } from './index.fixture';

function encode(event: object): string {
  return Object.keys(event)
    .map((key) => event[key])
    .join('\t');
}

describe('transform', () => {
  it('should be transformed', () => {
    expect(transform(encode(event))).toMatchSnapshot();
  });

  it('should fail on fields number', () => {
    expect(() => transform(encode({ a: '1', b: '2' }))).toThrow('Wrong event fields number.');
  });

  it('should fail on one field', () => {
    expect(() => transform(encode({ ...event, tr_tax_base: 'bad_tax_base' }))).toThrow(
      "Invalid value for field 'tr_tax_base'.",
    );
  });

  it('should fail on multiple fields', () => {
    expect(() =>
      transform(
        encode({
          ...event,
          dvce_ismobile: 'bad_dvce_ismobile',
          tr_tax_base: 'bad_tax_base',
        }),
      ),
    ).toThrow(["Invalid value for field 'dvce_ismobile'.", "Invalid value for field 'tr_tax_base'."].join('\n'));
  });

  it('should not return geo_location', () => {
    expect(() => transform(encode({ ...event, geo_latitude: '', geo_longitude: '' }))).not.toHaveProperty(
      'geo_location',
    );
  });
});
