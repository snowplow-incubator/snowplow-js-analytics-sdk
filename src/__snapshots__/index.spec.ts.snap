// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should be transformed 1`] = `
Object {
  "app_id": "angry-birds",
  "br_features_flash": false,
  "br_features_pdf": true,
  "collector_tstamp": "2013-11-26T00:02:05Z",
  "contexts_com_snowplowanalytics_snowplow_ua_parser_context_1": Array [
    Object {
      "deviceFamily": "Other",
      "osFamily": "Windows XP",
      "osMajor": null,
      "osMinor": null,
      "osPatch": null,
      "osPatchMinor": null,
      "osVersion": "Windows XP",
      "useragentFamily": "IE",
      "useragentMajor": "7",
      "useragentMinor": "0",
      "useragentPatch": null,
      "useragentVersion": "IE 7.0",
    },
  ],
  "contexts_org_schema_web_page_1": Array [
    Object {
      "author": "Fred Blundun",
      "breadcrumb": Array [
        "blog",
        "releases",
      ],
      "datePublished": "2014-11-06T00:00:00Z",
      "genre": "blog",
      "inLanguage": "en-US",
      "keywords": Array [
        "snowplow",
        "javascript",
        "tracker",
        "event",
      ],
    },
  ],
  "contexts_org_w3_performance_timing_1": Array [
    Object {
      "connectEnd": 1415358090183,
      "connectStart": 1415358090103,
      "domComplete": 0,
      "domContentLoadedEventEnd": 1415358091309,
      "domContentLoadedEventStart": 1415358090968,
      "domInteractive": 1415358090886,
      "domLoading": 1415358090270,
      "domainLookupEnd": 1415358090102,
      "domainLookupStart": 1415358090102,
      "fetchStart": 1415358089870,
      "loadEventEnd": 0,
      "loadEventStart": 0,
      "navigationStart": 1415358089861,
      "redirectEnd": 0,
      "redirectStart": 0,
      "requestStart": 1415358090183,
      "responseEnd": 1415358090265,
      "responseStart": 1415358090265,
      "unloadEventEnd": 1415358090287,
      "unloadEventStart": 1415358090270,
    },
  ],
  "derived_tstamp": "2013-11-26T00:03:57.886Z",
  "domain_sessionid": "2b15e5c8-d3b1-11e4-b9d6-1681e6b88ec1",
  "domain_sessionidx": 3,
  "domain_userid": "bc2e92ec6c204a14",
  "dvce_created_tstamp": "2013-11-26T00:03:57.885Z",
  "etl_tstamp": "2017-01-26T00:01:25.292Z",
  "event": "page_view",
  "event_fingerprint": "e3dbfa9cca0412c3d4052863cefb547f",
  "event_format": "jsonschema",
  "event_id": "c6ef3124-b53a-4b13-a233-0088f79dcbcb",
  "event_name": "link_click",
  "event_vendor": "com.snowplowanalytics.snowplow",
  "event_version": "1-0-0",
  "geo_city": "New York",
  "geo_country": "US",
  "geo_latitude": 37.443604,
  "geo_location": "37.443604,-122.4124",
  "geo_longitude": -122.4124,
  "geo_region": "TX",
  "geo_region_name": "Florida",
  "geo_zipcode": "94109",
  "ip_domain": "nuvox.net",
  "ip_isp": "FDN Communications",
  "ip_netspeed": "Cable/DSL",
  "ip_organization": "Bouygues Telecom",
  "name_tracker": "cloudfront-1",
  "network_userid": "ecdff4d0-9175-40ac-a8bb-325c49733607",
  "page_title": "On Analytics",
  "page_url": "http://www.snowplowanalytics.com",
  "page_urlfragment": "4-conclusion",
  "page_urlhost": "www.snowplowanalytics.com",
  "page_urlpath": "/product/index.html",
  "page_urlport": 80,
  "page_urlquery": "id=GTM-DLRG",
  "page_urlscheme": "http",
  "platform": "web",
  "true_tstamp": "2013-11-26T00:03:57.886Z",
  "txn_id": 41828,
  "unstruct_event_com_snowplowanalytics_snowplow_link_click_1": Object {
    "elementClasses": Array [
      "foreground",
    ],
    "elementId": "exampleLink",
    "targetUrl": "http://www.example.com",
  },
  "user_fingerprint": "2161814971",
  "user_id": "jon.doe@email.com",
  "user_ipaddress": "92.231.54.234",
  "v_collector": "clj-tomcat-0.1.0",
  "v_etl": "serde-0.5.2",
  "v_tracker": "js-2.1.0",
}
`;
